File: windows/MyWeChat.Windows/Utils/EncryptionService.cs
using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;

namespace MyWeChat.Windows.Utils
{
    /// <summary>
    /// 加密服务（AES-256-GCM）
    /// 提供字符串和字节数组的加密/解密功能
    /// 支持两种密钥模式：
    /// 1. 本地密钥（用于日志加密，基于机器特征）
    /// 2. 会话密钥（用于通讯加密，从服务器交换获得）
    /// </summary>
    public static class EncryptionService
    {
        // 会话密钥（用于通讯加密）
        private static byte[]? _sessionKey;
        private static readonly object _sessionKeyLock = new object();

        /// <summary>
        /// 设置会话密钥（用于通讯加密）
        /// </summary>
        public static void SetSessionKey(byte[] sessionKey)
        {
            if (sessionKey == null || sessionKey.Length != 32)
            {
                throw new ArgumentException("会话密钥必须是32字节", nameof(sessionKey));
            }

            lock (_sessionKeyLock)
            {
                _sessionKey = new byte[32];
                Buffer.BlockCopy(sessionKey, 0, _sessionKey, 0, 32);
                Logger.LogInfo("会话密钥已设置（用于通讯加密）");
            }
        }

        /// <summary>
        /// 清除会话密钥
        /// </summary>
        public static void ClearSessionKey()
        {
            lock (_sessionKeyLock)
            {
                if (_sessionKey != null)
                {
                    Array.Clear(_sessionKey, 0, _sessionKey.Length);
                    _sessionKey = null;
                }
            }
        }

        /// <summary>
        /// 检查会话密钥是否已设置
        /// </summary>
        public static bool HasSessionKey()
        {
            lock (_sessionKeyLock)
            {
                return _sessionKey != null;
            }
        }

        /// <summary>
        /// 加密字符串（用于通讯，使用会话密钥）
        /// </summary>
        public static string EncryptStringForCommunication(string plainText)
        {
            if (string.IsNullOrEmpty(plainText))
            {
                return string.Empty;
            }

            try
            {
                byte[] plainBytes = Encoding.UTF8.GetBytes(plainText);
                byte[] encrypted = EncryptBytesWithKey(plainBytes, GetSessionKey());
                return Convert.ToBase64String(encrypted);
            }
            catch (Exception ex)
            {
                Logger.LogError($"加密通讯字符串失败: {ex.Message}", ex);
                throw;
            }
        }

        /// <summary>
        /// 解密字符串（用于通讯，使用会话密钥）
        /// </summary>
        public static string DecryptStringForCommunication(string cipherText)
        {
            if (string.IsNullOrEmpty(cipherText))
            {
                return string.Empty;
            }

            try
            {
                byte[] cipherBytes = Convert.FromBase64String(cipherText);
                byte[] decrypted = DecryptBytesWithKey(cipherBytes, GetSessionKey());
                return Encoding.UTF8.GetString(decrypted);
            }
            catch (Exception ex)
            {
                Logger.LogError($"解密通讯字符串失败: {ex.Message}", ex);
                throw;
            }
        }

        /// <summary>
        /// 解密字符串（用于HTTP API，直接使用指定的HTTP会话密钥）
        /// </summary>
        public static string DecryptStringForHttp(byte[] httpSessionKey, string cipherText)
        {
            if (string.IsNullOrEmpty(cipherText))
            {
                return string.Empty;
            }

            if (httpSessionKey == null || httpSessionKey.Length != 32)
            {
                throw new ArgumentException("HTTP会话密钥必须是32字节", nameof(httpSessionKey));
            }

            try
            {
                byte[] cipherBytes = Convert.FromBase64String(cipherText);
                byte[] decrypted = DecryptBytesWithKey(cipherBytes, httpSessionKey);
                return Encoding.UTF8.GetString(decrypted);
            }
            catch (Exception ex)
            {
                Logger.LogError($"解密HTTP字符串失败: {ex.Message}", ex);
                throw;
            }
        }

        /// <summary>
        /// 加密字符串（用于日志，使用本地密钥）
        /// </summary>
        public static string EncryptStringForLog(string plainText)
        {
            if (string.IsNullOrEmpty(plainText))
            {
                return string.Empty;
            }

            try
            {
                byte[] plainBytes = Encoding.UTF8.GetBytes(plainText);
                byte[] encrypted = EncryptBytesWithKey(plainBytes, KeyDerivationService.GetEncryptionKey());
                return Convert.ToBase64String(encrypted);
            }
            catch (Exception ex)
            {
                Logger.LogError($"加密日志字符串失败: {ex.Message}", ex);
                throw;
            }
        }

        /// <summary>
        /// 解密字符串（用于日志，使用本地密钥）
        /// </summary>
        public static string DecryptStringForLog(string cipherText)
        {
            if (string.IsNullOrEmpty(cipherText))
            {
                return string.Empty;
            }

            try
            {
                byte[] cipherBytes = Convert.FromBase64String(cipherText);
                byte[] decrypted = DecryptBytesWithKey(cipherBytes, KeyDerivationService.GetEncryptionKey());
                return Encoding.UTF8.GetString(decrypted);
            }
            catch (Exception ex)
            {
                Logger.LogError($"解密日志字符串失败: {ex.Message}", ex);
                throw;
            }
        }

        /// <summary>
        /// 获取会话密钥
        /// </summary>
        private static byte[] GetSessionKey()
        {
            lock (_sessionKeyLock)
            {
                if (_sessionKey == null)
                {
                    throw new InvalidOperationException("会话密钥未设置，无法进行通讯加密");
                }
                return _sessionKey;
            }
        }

        /// <summary>
        /// 使用指定密钥加密字节数组
        /// 格式：nonce(12字节) + ciphertext + tag(16字节)
        /// </summary>
        private static byte[] EncryptBytesWithKey(byte[] plainBytes, byte[] key)
        {
            if (plainBytes == null || plainBytes.Length == 0)
            {
                return Array.Empty<byte>();
            }

            if (key == null || key.Length != 32)
            {
                throw new ArgumentException("密钥必须是32字节", nameof(key));
            }

            try
            {
                // 使用新的构造函数指定 tag 大小（16字节）
                using (AesGcm aesGcm = new AesGcm(key, 16))
                {
                    // 生成随机 nonce（12字节）
                    byte[] nonce = new byte[12];
                    using (RandomNumberGenerator rng = RandomNumberGenerator.Create())
                    {
                        rng.GetBytes(nonce);
                    }

                    // 加密
                    byte[] ciphertext = new byte[plainBytes.Length];
                    byte[] tag = new byte[16];
                    aesGcm.Encrypt(nonce, plainBytes, ciphertext, tag);

                    // 组合：nonce + ciphertext + tag
                    byte[] result = new byte[12 + ciphertext.Length + 16];
                    Buffer.BlockCopy(nonce, 0, result, 0, 12);
                    Buffer.BlockCopy(ciphertext, 0, result, 12, ciphertext.Length);
                    Buffer.BlockCopy(tag, 0, result, 12 + ciphertext.Length, 16);

                    return result;
                }
            }
            catch (Exception ex)
            {
                Logger.LogError($"加密字节数组失败: {ex.Message}", ex);
                throw;
            }
        }

        /// <summary>
        /// 使用指定密钥解密字节数组
        /// 格式：nonce(12字节) + ciphertext + tag(16字节)
        /// </summary>
        private static byte[] DecryptBytesWithKey(byte[] cipherBytes, byte[] key)
        {
            if (cipherBytes == null || cipherBytes.Length < 28) // 至少需要 12(nonce) + 0(ciphertext) + 16(tag)
            {
                return Array.Empty<byte>();
            }

            if (key == null || key.Length != 32)
            {
                throw new ArgumentException("密钥必须是32字节", nameof(key));
            }

            try
            {
                // 提取 nonce、ciphertext 和 tag
                byte[] nonce = new byte[12];
                Buffer.BlockCopy(cipherBytes, 0, nonce, 0, 12);

                int ciphertextLength = cipherBytes.Length - 28; // 总长度 - nonce(12) - tag(16)
                byte[] ciphertext = new byte[ciphertextLength];
                Buffer.BlockCopy(cipherBytes, 12, ciphertext, 0, ciphertextLength);

                byte[] tag = new byte[16];
                Buffer.BlockCopy(cipherBytes, 12 + ciphertextLength, tag, 0, 16);

                // 解密（使用新的构造函数指定 tag 大小）
                using (AesGcm aesGcm = new AesGcm(key, 16))
                {
                    byte[] plaintext = new byte[ciphertextLength];
                    aesGcm.Decrypt(nonce, ciphertext, tag, plaintext);
                    return plaintext;
                }
            }
            catch (Exception ex)
            {
                Logger.LogError($"解密字节数组失败: {ex.Message}", ex);
                throw;
            }
        }

        // 为了向后兼容，保留旧的方法（使用本地密钥）
        /// <summary>
        /// 加密字符串（使用本地密钥，用于日志）
        /// </summary>
        [Obsolete("请使用 EncryptStringForLog 或 EncryptStringForCommunication")]
        public static string EncryptString(string plainText)
        {
            return EncryptStringForLog(plainText);
        }

        /// <summary>
        /// 解密字符串（使用本地密钥，用于日志）
        /// </summary>
        [Obsolete("请使用 DecryptStringForLog 或 DecryptStringForCommunication")]
        public static string DecryptString(string cipherText)
        {
            return DecryptStringForLog(cipherText);
        }

        /// <summary>
        /// 加密字节数组（使用本地密钥，用于日志）
        /// </summary>
        [Obsolete("请使用 EncryptBytesWithKey")]
        public static byte[] EncryptBytes(byte[] plainBytes)
        {
            return EncryptBytesWithKey(plainBytes, KeyDerivationService.GetEncryptionKey());
        }

        /// <summary>
        /// 解密字节数组（使用本地密钥，用于日志）
        /// </summary>
        [Obsolete("请使用 DecryptBytesWithKey")]
        public static byte[] DecryptBytes(byte[] cipherBytes)
        {
            return DecryptBytesWithKey(cipherBytes, KeyDerivationService.GetEncryptionKey());
        }
    }
}

File: server/app/utils/encryption_service.py
"""
加密服务（AES-256-GCM）
与客户端 C# 版本兼容
支持会话密钥（用于通讯加密）和本地密钥（用于日志加密）分离
"""
import os
import base64
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.backends import default_backend
from typing import Optional, Dict


class EncryptionService:
    """加密服务（AES-256-GCM）"""
    
    _instance = None
    _local_key = None  # 本地密钥（用于日志加密）
    _session_keys: Dict[str, bytes] = {}  # 会话密钥字典（WebSocket连接ID -> 会话密钥）
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(EncryptionService, cls).__new__(cls)
        return cls._instance
    
    def __init__(self):
        if self._local_key is None:
            self._local_key = self._get_local_encryption_key()
    
    def _get_local_encryption_key(self) -> bytes:
        """获取本地加密密钥（32字节，256位，用于日志加密）"""
        # 优先从环境变量读取密钥
        env_key = os.getenv("MYWECHAT_ENCRYPTION_KEY")
        if env_key:
            # 如果环境变量是 base64 编码的，先解码
            try:
                key = base64.b64decode(env_key)
                if len(key) == 32:
                    return key
            except:
                pass
        
        # 如果没有环境变量，使用默认密钥（基于固定盐值）
        # 注意：生产环境应该使用环境变量设置密钥
        default_salt = b"MyWeChat_Server_Encryption_Salt_2024"
        default_password = b"MyWeChat_Default_Password_Change_In_Production"
        
        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=32,
            salt=default_salt,
            iterations=100000,
            backend=default_backend()
        )
        key = kdf.derive(default_password)
        return key
    
    def set_session_key(self, connection_id: str, session_key: bytes):
        """设置会话密钥（用于通讯加密）"""
        if len(session_key) != 32:
            raise ValueError("会话密钥必须是32字节")
        self._session_keys[connection_id] = session_key
        print(f"会话密钥已设置（连接ID: {connection_id[:8]}...）")
    
    def get_session_key(self, connection_id: str) -> Optional[bytes]:
        """获取会话密钥（用于通讯加密）"""
        return self._session_keys.get(connection_id)
    
    def remove_session_key(self, connection_id: str):
        """移除会话密钥"""
        if connection_id in self._session_keys:
            del self._session_keys[connection_id]
            print(f"会话密钥已移除（连接ID: {connection_id[:8]}...）")
    
    def has_session_key(self, connection_id: str) -> bool:
        """检查是否有会话密钥"""
        return connection_id in self._session_keys
    
    def encrypt_string_for_communication(self, connection_id: str, plain_text: str) -> str:
        """加密字符串（用于通讯，使用会话密钥）"""
        if not plain_text:
            return ""
        
        session_key = self.get_session_key(connection_id)
        if session_key is None:
            raise ValueError(f"连接 {connection_id} 的会话密钥未设置")
        
        try:
            plain_bytes = plain_text.encode('utf-8')
            encrypted = self._encrypt_bytes_with_key(plain_bytes, session_key)
            return base64.b64encode(encrypted).decode('utf-8')
        except Exception as e:
            print(f"加密通讯字符串失败: {e}")
            raise
    
    def decrypt_string_for_communication(self, connection_id: str, cipher_text: str) -> str:
        """解密字符串（用于通讯，使用会话密钥）"""
        if not cipher_text:
            return ""
        
        session_key = self.get_session_key(connection_id)
        if session_key is None:
            raise ValueError(f"连接 {connection_id} 的会话密钥未设置")
        
        try:
            cipher_bytes = base64.b64decode(cipher_text)
            decrypted = self._decrypt_bytes_with_key(cipher_bytes, session_key)
            return decrypted.decode('utf-8')
        except Exception as e:
            print(f"解密通讯字符串失败: {e}")
            raise
    
    def encrypt_string_for_log(self, plain_text: str) -> str:
        """加密字符串（用于日志，使用本地密钥）"""
        if not plain_text:
            return ""
        
        try:
            plain_bytes = plain_text.encode('utf-8')
            encrypted = self._encrypt_bytes_with_key(plain_bytes, self._local_key)
            return base64.b64encode(encrypted).decode('utf-8')
        except Exception as e:
            print(f"加密日志字符串失败: {e}")
            raise
    
    def decrypt_string_for_log(self, cipher_text: str) -> str:
        """解密字符串（用于日志，使用本地密钥）"""
        if not cipher_text:
            return ""
        
        try:
            cipher_bytes = base64.b64decode(cipher_text)
            decrypted = self._decrypt_bytes_with_key(cipher_bytes, self._local_key)
            return decrypted.decode('utf-8')
        except Exception as e:
            print(f"解密日志字符串失败: {e}")
            raise
    
    def _encrypt_bytes_with_key(self, plain_bytes: bytes, key: bytes) -> bytes:
        """使用指定密钥加密字节数组
        格式：nonce(12字节) + ciphertext + tag(16字节)
        """
        if not plain_bytes:
            return b""
        
        try:
            # 生成随机 nonce（12字节）
            import secrets
            nonce = secrets.token_bytes(12)
            
            # 加密
            aesgcm = AESGCM(key)
            ciphertext = aesgcm.encrypt(nonce, plain_bytes, None)
            
            # 组合：nonce + ciphertext（ciphertext 已经包含 tag）
            # 注意：AESGCM.encrypt 返回的是 ciphertext + tag
            # 所以格式是：nonce(12) + ciphertext + tag(16)
            result = nonce + ciphertext
            return result
        except Exception as e:
            print(f"加密字节数组失败: {e}")
            raise
    
    def _decrypt_bytes_with_key(self, cipher_bytes: bytes, key: bytes) -> bytes:
        """使用指定密钥解密字节数组
        格式：nonce(12字节) + ciphertext + tag(16字节)
        """
        if not cipher_bytes or len(cipher_bytes) < 28:  # 至少需要 12(nonce) + 0(ciphertext) + 16(tag)
            return b""
        
        try:
            # 提取 nonce、ciphertext 和 tag
            nonce = cipher_bytes[:12]
            ciphertext_with_tag = cipher_bytes[12:]
            
            # 解密
            aesgcm = AESGCM(key)
            plaintext = aesgcm.decrypt(nonce, ciphertext_with_tag, None)
            return plaintext
        except Exception as e:
            print(f"解密字节数组失败: {e}")
            raise
    
    # 为了向后兼容，保留旧的方法（使用本地密钥）
    def encrypt_string(self, plain_text: str) -> str:
        """加密字符串（使用本地密钥，用于日志）"""
        return self.encrypt_string_for_log(plain_text)
    
    def decrypt_string(self, cipher_text: str) -> str:
        """解密字符串（使用本地密钥，用于日志）"""
        return self.decrypt_string_for_log(cipher_text)


# 全局加密服务实例
encryption_service = EncryptionService()

